torch/csrc/jit/passes/autocast.cpp

```diff
@@ -0,0 +1,456 @@
+
+#include <torch/csrc/jit/passes/autocast.h>
+
+#include <ATen/autocast_mode.h>
+#include <c10/core/ScalarType.h>
+#include <c10/util/Exception.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/cuda_graph_fuser.h>
+#include <torch/csrc/jit/passes/quantization/helper.h>
+
+#include <stack>
+#include <unordered_set>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+// TODO: Turn on autocast by default. default turned off to avoid tests failures
+// as we prototype the support
+bool autocast_enabled = false;
+
+struct AutocastContext {
+  bool enabled = false;
+  bool cpu_enabled = false;
+  c10::ScalarType scalar_type = c10::ScalarType::Undefined;
+  c10::ScalarType cpu_scalar_type = c10::ScalarType::Undefined;
+
+  operator bool() const {
+    return enabled || cpu_enabled;
+  }
+};
+
+struct AutocastScope {
+  Value* instance = nullptr;
+  AutocastContext context;
+  void stack(const AutocastContext& parent_context) {}
+};
+
+bool isAutocastNode(Value* value) {
+  const auto class_name = getModuleName(value);
+  return class_name.has_value() &&
+      (*class_name == "__torch__.torch.cuda.amp.autocast_mode.autocast" ||
+       *class_name == "__torch__.torch.cpu.amp.autocast_mode.autocast" ||
+       *class_name == "__torch__.torch.autocast_mode.autocast");
+}
+
+// If we have an autocast instance, return it
+//
+// This is the pattern we're looking for (this is done after
+//  autocast.__init__() has been inlined)
+//
+// %4 : bool = prim::Constant[value=1]()
+// %5 : __torch__.torch.cuda.amp.autocast_mode.autocast = prim::CreateObject()
+//  = prim::SetAttr[name="_enabled"](%5, %4)
```

Comment:
This isnt necessarily needed now, but in the future we would like to apply optimizations to optimize out object construction / accesses (Scalar Replacement of Aggregates), so eventually it would be nice to figure out how to embed this in IR (perhaps relevant re: to actually invoking `torch.set_autocast_cpu_enabled`